core/process.py

```diff
@@ -928,6 +949,10 @@ def StartProcess(self, why):
     # Program invariant: We keep track of every child process!
     self.job_state.AddChildProcess(pid, self)
 
+    # QUESTION: Why do we need to add the last PID?
```

Comment:
We only need this to make sure we set the correct exit status in [Pipeline.WhenDone()](https://github.com/oilshell/oil/blob/master/core/process.py#L1184-L1198). We could relax the check in that code to just always write to the last status slot if `Pipeline.WhenDone()` gets called with a `PID` that isn't part of `Pipeline.pids`

Reply:
If we did this refactor and used the last part of the pipeline as the leader of the pipeline's group (instead of the first) we could probably get rid of `fg_pipeline`! Let me give that a try

Reply:
Hmm there's actually a trade-off here. By removing `fg_pipeline` we would muddy the logic in `RunPipeline()` somewhat. It would have to conditionally start the last part of the pipeline first if it would fork or run it after starting the other parts if it's just a builtin, etc... Awkward logic is probably better than extra state, though. Let's see how it ends up looking....